--- a/tools/c_rehash.in	2022-09-28 18:39:43.347059893 +0200
+++ b/tools/c_rehash.in	2022-09-28 09:01:33.314213892 +0200
@@ -140,6 +146,7 @@ sub hash_dir {
             next;
         }
         link_hash_cert($fname) if ($cert);
+        link_hash_cert_old($fname) if ($cert);
         link_hash_crl($fname) if ($crl);
     }
 
@@ -192,22 +200,26 @@ sub compute_hash {
 # certificate fingerprints
 
 sub link_hash_cert {
-    link_hash($_[0], 'cert');
+    link_hash($_[0], 'cert', '-subject_hash');
+}
+
+sub link_hash_cert_old {
+    link_hash($_[0], 'cert', '-subject_hash_old');
 }
 
 # Same as above except for a CRL. CRL links are of the form <hash>.r<n>
 
 sub link_hash_crl {
-    link_hash($_[0], 'crl');
+    link_hash($_[0], 'crl', '-hash');
 }
 
 sub link_hash {
-    my ($fname, $type) = @_;
+    my ($fname, $type, $hash_type) = @_;
     my $is_cert = $type eq 'cert';
 
     my ($hash, $fprint) = compute_hash($openssl,
                                        $is_cert ? "x509" : "crl",
-                                       $is_cert ? $x509hash : $crlhash,
+                                       $hash_type,
                                        "-fingerprint", "-noout",
                                        "-in", $fname);
     chomp $hash;
