---
 tools/c_rehash.in |   12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

--- a/tools/c_rehash.in
+++ b/tools/c_rehash.in
@@ -136,8 +136,11 @@ sub hash_dir {
         # Check to see if certificates and/or CRLs present.
         my ($cert, $crl) = check_file($fname);
         if (!$cert && !$crl) {
-            print STDERR "WARNING: $fname does not contain a certificate or CRL: skipping\n";
-            next;
+            ($cert, $crl) = check_file("$openssl x509 -in \"$fname\" -inform der  -outform pem | ");
+            if (!$cert && !$crl) {
+                print STDERR "WARNING: $fname does not contain a certificate or CRL: skipping\n";
+                next;
+            }
         }
         link_hash_cert($fname) if ($cert);
         link_hash_cert_old($fname) if ($cert);
@@ -215,6 +218,14 @@ sub link_hash {
                                        $hash_type,
                                        "-fingerprint", "-noout",
                                        "-in", $fname);
+    if (!$hash || !fprint) {
+        my ($hash, $fprint) = compute_hash($openssl,
+                                           $is_cert ? "x509" : "crl",
+                                           $hash_type,
+                                           "-fingerprint", "-noout",
+                                           "-in", $fname,
+                                           "-inform der");
+    }
     chomp $hash;
     chomp $fprint;
     return if !$hash;
